### PR TITLE
Change implementation of fnearest()

### PR DESCRIPTION
**Pull request by chfast**
This implementation does not need to modify rounding direction to work
correctly under any rounding direction.

### CODE REVIEW

File: lib/fizzy/execute.cpp [hunk 1]
```diff
@@ -590,34 +589,13 @@ T fnearest(T value) noexcept
     if (std::isnan(value))
```
Comment (axic) [inline lib/fizzy/execute.cpp:589]: The comment above is invalid now.

File: lib/fizzy/execute.cpp [hunk 2]
```diff
@@ -590,34 +589,13 @@ T fnearest(T value) noexcept
...
-    // volatile is used to prevent compiler from moving the last std::fesetround() before.
-    const volatile auto result = std::nearbyint(value);
-
-    // Reset the rounding direction if we know the previous value and it was actually modified.
-    if (current_rounding_direction >= 0 && current_rounding_direction != FE_TONEAREST)
-        std::fesetround(current_rounding_direction);
-
-    return result;
+    const auto t = std::trunc(value);
+    if (const auto diff = std::abs(value - t); diff > 0.5f || (diff == 0.5f && !is_even(t)))
```
Comment (axic) [inline lib/fizzy/execute.cpp:595]: Wouldn't it be less branching to have two if cases than using abs? Or is abs constexpr and the compiler can optimise all this out?

Comment (chfast) [reply lib/fizzy/execute.cpp:595]: The `abs` (or `fabs` in C) is recognized and replaced with `__builtin_abs` or something. This just masks out the sign bit: `x & 0x7fffffff`.

**Comment by chfast** [comment]
I also have another implementation based on bit manipulation which looks to be faster but I did not perform proper benchmarks yet.
I decided to start with this one, because it does not need special utils like `class FP`.

We may wait with that for the next release until having more FP test cases.

File: lib/fizzy/execute.cpp [hunk 3]
```diff
@@ -585,39 +584,20 @@ T fnearest(T value) noexcept
...
     if (std::isnan(value))
         return std::numeric_limits<T>::quiet_NaN();  // Positive canonical NaN.
 
-    // This implementation uses std::nearbyint() which rounds to integer using current rounding
-    // direction. The rounding direction is temporarily set to "to nearest" for the case when
-    // the current rounding direction is different.
+    // Check if the input integer (as floating-point type) is even.
+    // There is a faster way of doing that by bit manipulations of mantissa and exponent,
+    // but this one is compact and this function is rarely called.
+    static constexpr auto is_even = [](auto i) noexcept { return std::fmod(i, 2.0f) == 0; };
```
Comment (gumb0) [inline lib/fizzy/execute.cpp:593]: `std::fmod` here can be one of two overloaded functions depending on `T`, right?

Maybe for less confusion
```suggestion
    static constexpr auto is_even = [](auto i) noexcept { return std::fmod(i, T{2.0}) == 0; };
```
but not sure if it's better

Comment (chfast) [reply lib/fizzy/execute.cpp]: Changed to `T{2}` and `T{0}`. The usage of `{}` guarantees lossless conversion.

File: lib/fizzy/execute.cpp [hunk 4]
```diff
@@ -585,39 +584,21 @@ T fnearest(T value) noexcept
...
-    const volatile auto result = std::nearbyint(value);
-
-    // Reset the rounding direction if we know the previous value and it was actually modified.
-    if (current_rounding_direction >= 0 && current_rounding_direction != FE_TONEAREST)
-        std::fesetround(current_rounding_direction);
-
-    return result;
+    // This implementation is based on adjusting the result produced by trunc() by +-1 when needed.
+    const auto t = std::trunc(value);
+    if (const auto diff = std::abs(value - t); diff > 0.5f || (diff == 0.5f && !is_even(t)))
```
Comment (axic) [inline lib/fizzy/execute.cpp:598]: Hm, is there any conversion involved if `diff` is double and we check against `0.5f`?

Comment (axic) [reply lib/fizzy/execute.cpp:598]: So should `T{0.5f}` be better?

Comment (chfast) [reply lib/fizzy/execute.cpp:598]: The convention is `0.5f` is promoted to `0.5`.

Comment (axic) [reply lib/fizzy/execute.cpp:598]: Right, but is that at compile time?

Comment (chfast) [reply lib/fizzy/execute.cpp:598]: In practice, yes. I will use `T{0.5}`

**Comment by chfast** [comment]
Tested with #511: https://app.circleci.com/pipelines/github/wasmx/fizzy/4401/workflows/c600c27b-1f5e-4d6e-a2db-71613aa947d5

**Review by gumb0 — APPROVED** [review]

**Comment by chfast** [comment]
I'm checking some additional options because there are some performance regressions.

**Comment by chfast** [comment]
GCC:
```
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     -0.0166         -0.0166            86            85            86            85
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    -0.0143         -0.0143          1308          1289          1308          1289
fizzy/execute/ecpairing/onepoint_mean                             -0.0481         -0.0481        439794        418619        439798        418623
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   -0.0126         -0.0126           104           102           104           102
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  -0.0132         -0.0132          1520          1500          1520          1500
fizzy/execute/memset/256_bytes_mean                               -0.0312         -0.0312             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             -0.0294         -0.0294          1576          1529          1576          1529
fizzy/execute/mul256_opt0/input0_mean                             -0.0820         -0.0820            28            26            28            26
fizzy/execute/mul256_opt0/input1_mean                             -0.0858         -0.0858            29            26            29            26
fizzy/execute/ramanujan_pi/33_runs_mean                           -0.0174         -0.0174           135           132           135           132
fizzy/execute/sha1/512_bytes_rounds_1_mean                        -0.0021         -0.0021            93            93            93            93
fizzy/execute/sha1/512_bytes_rounds_16_mean                       +0.0003         +0.0004          1297          1297          1297          1297
fizzy/execute/sha256/512_bytes_rounds_1_mean                      -0.0058         -0.0058            94            94            94            94
fizzy/execute/sha256/512_bytes_rounds_16_mean                     -0.0018         -0.0018          1297          1295          1297          1295
fizzy/execute/taylor_pi/pi_1000000_runs_mean                      -0.0665         -0.0665         42825         39978         42825         39978
fizzy/execute/micro/eli_interpreter/halt_mean                     +0.0229         +0.0229             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  +0.0775         +0.0775             5             5             5             5
fizzy/execute/micro/factorial/10_mean                             +0.0131         +0.0131             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             +0.0096         +0.0097             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             +0.0060         +0.0060          7502          7547          7502          7547
fizzy/execute/micro/host_adler32/1_mean                           +0.0236         +0.0236             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         +0.0102         +0.0102             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        -0.0106         -0.0106            30            29            30            29
fizzy/execute/micro/spinner/1_mean                                +0.0115         +0.0115             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             -0.0243         -0.0243            10            10            10            10
```

GCC+LTO:
```
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     -0.0146         -0.0146            84            83            84            83
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    -0.0174         -0.0174          1277          1255          1277          1255
fizzy/execute/ecpairing/onepoint_mean                             +0.0066         +0.0066        424685        427480        424689        427484
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   +0.0286         +0.0286            98           101            98           101
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  +0.0227         +0.0227          1431          1464          1431          1464
fizzy/execute/memset/256_bytes_mean                               +0.0057         +0.0057             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             +0.0091         +0.0091          1544          1558          1544          1558
fizzy/execute/mul256_opt0/input0_mean                             +0.0105         +0.0105            26            26            26            26
fizzy/execute/mul256_opt0/input1_mean                             +0.0134         +0.0134            26            26            26            26
fizzy/execute/ramanujan_pi/33_runs_mean                           +0.0385         +0.0385           126           131           126           131
fizzy/execute/sha1/512_bytes_rounds_1_mean                        +0.0085         +0.0085            90            91            90            91
fizzy/execute/sha1/512_bytes_rounds_16_mean                       +0.0118         +0.0118          1252          1267          1252          1267
fizzy/execute/sha256/512_bytes_rounds_1_mean                      +0.0138         +0.0138            91            92            91            92
fizzy/execute/sha256/512_bytes_rounds_16_mean                     +0.0116         +0.0116          1252          1267          1252          1267
fizzy/execute/taylor_pi/pi_1000000_runs_mean                      -0.0223         -0.0223         42176         41237         42177         41238
fizzy/execute/micro/eli_interpreter/halt_mean                     -0.0304         -0.0304             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  +0.0399         +0.0399             5             5             5             5
fizzy/execute/micro/factorial/10_mean                             -0.0135         -0.0135             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             -0.0191         -0.0191             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             +0.0026         +0.0026          7488          7507          7488          7507
fizzy/execute/micro/host_adler32/1_mean                           -0.0062         -0.0062             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         +0.0285         +0.0285             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        -0.0057         -0.0057            30            30            30            30
fizzy/execute/micro/spinner/1_mean                                +0.0172         +0.0172             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             +0.0434         +0.0434            10            11            10            11
```

So not really significant difference (none of the benchmarks uses this instruction). The change is not driven by performance anyway.

When marking the function with `noinline` and `cold` attributes, it gives some performance improvement, but I'm leaving that for some other time.

**Review by axic — APPROVED** [review]

**Pull request merged**